### PR TITLE
feat(pacquet): `pack`

### PR DESCRIPTION
**Pull request by KSXGitHub**
## Summary by CodeRabbit

* **New Features**
  * Added a `pack` command to generate publishable tarballs from a project, including recursive workspace packing.
  * Added dry-run, JSON output, output-path templating, gzip level selection, and optional README embedding.
  * Introduced exportable/publish-ready manifest creation (obfuscation, workspace/catalog/jsr spec rewriting, and `publishConfig` hoisting).
* **Bug Fixes**
  * Improved manifest normalization/validation and deterministic tarball generation/unpacked-size reporting.
  * Hardened `main`/`bin` handling to prevent path traversal and tightened file/symlink inclusion behavior.
  * Refined lifecycle/script execution and `publishConfig.directory` behavior.
* **Documentation**
  * Expanded `pack` CLI documentation and options.
* **Tests**
  * Added coverage for packing, manifest creation/transformation, output formatting, and traversal/symlink edge cases.

### CODE REVIEW

File: pacquet/crates/cli/src/cli_args/pack.rs [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    pub json: bool,
+
+    /// Customize the output path. `%s` expands to the package name and
+    /// `%v` to the version, e.g. `%s.tgz` or `some-dir/%s-%v.tgz`.
+    #[clap(long)]
+    pub out: Option<String>,
+
+    /// gzip compression level (`0`–`9`) for the tarball.
+    #[clap(long = "pack-gzip-level")]
+    pub pack_gzip_level: Option<u32>,
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pacquet/crates/cli/src/cli_args/pack.rs:59]: Fixed in f81941b94e — added `value_parser = clap::value_parser!(u32).range(0..=9)` on `pack_gzip_level`, so out-of-range values now fail at argument-parsing time.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pacquet/crates/cli/src/cli_args/pack.rs [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        let mut packed: Vec<PackResultJson> = Vec::new();
+        for chunk in &chunks {
+            for root in chunk {
+                let project = graph[root].package.project;
+                let manifest = project.manifest.value();
+                let has_name = manifest.get("name").and_then(|name| name.as_str()).is_some();
+                let has_version = manifest.get("version").and_then(|v| v.as_str()).is_some();
+                if !has_name || !has_version {
+                    continue;
+                }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pacquet/crates/cli/src/cli_args/pack.rs:146]: Fixed in f81941b94e — `has_name` / `has_version` now require a non-empty string (`.is_some_and(|s| !s.is_empty())`), so empty values are skipped to match the `pacquet_pack::api` precondition.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pacquet/crates/exportable-manifest/src/create.rs [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+        if let Some(deps) =
+            make_publish_dependencies(dir, original.get(field), opts, DependencyKind::Regular)?
+        {
+            publish.insert(field.to_string(), deps);
+        }
+    }
+    if let Some(peer) = original.get("peerDependencies") {
+        let converted = make_publish_dependencies(dir, Some(peer), opts, DependencyKind::Peer)?
+            .unwrap_or_else(|| Value::Object(Map::new()));
+        publish.insert("peerDependencies".to_string(), converted);
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pacquet/crates/exportable-manifest/src/create.rs]: Fixed in f81941b94e — `peerDependencies` now flows through the same `if let Some(converted) = make_publish_dependencies(...)` path as the other dependency fields, so a present-but-non-object value is left in place instead of being coerced to `{}`.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pacquet/crates/exportable-manifest/src/transform.rs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+/// Reject a manifest missing the registry-required `name` / `version`.
+/// A field present but not a non-empty string counts as missing, matching
+/// the JS `if (!manifest.name)` truthiness check.
+fn transform_required_fields(manifest: &Map<String, Value>) -> Result<(), TransformError> {
+    for field in ["name", "version"] {
+        let present =
+            manifest.get(field).and_then(Value::as_str).is_some_and(|value| !value.is_empty());
+        if !present {
+            return Err(TransformError::MissingRequiredField { field });
+        }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pacquet/crates/exportable-manifest/src/transform.rs:55]: Fixed in f81941b94e — `transform_required_fields` now gates on JavaScript truthiness via a new `is_truthy` helper, so a present non-string value (e.g. `"name": 1`) passes just like upstream's `if (!manifest.name)`. Only absent / `null` / `false` / `0` / `""` count as missing. Added tests covering empty-string and falsy rejection plus non-string acceptance.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pacquet/crates/pack/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,645 @@
...
+) -> Result<(), PackError> {
+    if node_linker == NodeLinker::Hoisted {
+        return Ok(());
+    }
+    for field in ["bundledDependencies", "bundleDependencies"] {
+        if manifest.get(field).is_some_and(|value| !value.is_null()) {
+            return Err(PackError::BundledDependenciesWithoutHoisted {
+                field,
+                node_linker: node_linker_str(node_linker),
+            });
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pacquet/crates/pack/src/lib.rs:436]: Fixed in f81941b94e — the guard now uses JS truthiness (`is_truthy`) instead of `!value.is_null()`, so `bundleDependencies: false` (and `bundledDependencies: false`) packs cleanly under the default non-hoisted linker. Added a regression test for the `false` case.

One deviation from the suggested diff: I used full JS truthiness rather than `matches!(value, Value::Bool(true) | Value::Array(_))`, because upstream's `if (bundledDependencies)` also rejects a non-empty string, a non-zero number, and `{}` — the narrower match would let those through, diverging from pnpm.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pacquet/crates/pack/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,645 @@
...
+    let Some(workspace_dir) = &opts.workspace_dir else { return };
+    if dir == workspace_dir || files.iter().any(|file| contains_license(file)) {
+        return;
+    }
+    let Ok(entries) = std::fs::read_dir(workspace_dir) else { return };
+    for entry in entries.flatten() {
+        let name = entry.file_name().to_string_lossy().into_owned();
+        if is_license_filename(&name) {
+            files_map.insert(format!("package/{name}"), workspace_dir.join(&name));
+        }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pacquet/crates/pack/src/lib.rs:629]: Fixed in f81941b94e — the workspace LICENSE injection now only inserts when the entry resolves to a regular file (`fs::metadata(entry.path()).is_ok_and(|m| m.is_file())`), so a directory named `LICENSE` no longer trips the later read/size pass. I used `fs::metadata` (which follows symlinks) rather than `entry.file_type()` so a symlinked license still counts, matching fast-glob's default symlink resolution + `onlyFiles`.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pacquet/crates/cli/src/cli_args/pack.rs [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+                    config,
+                    out.clone(),
+                    pack_destination.clone(),
+                );
+                let result = api::<Reporter, Host>(&options)
+                    .map_err(miette::Report::new)
+                    .wrap_err_with(|| format!("pack {}", project.root_dir.display()))?;
+                packed.push(to_pack_result_json(&result));
+            }
+        }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pacquet/crates/cli/src/cli_args/pack.rs:158]: Verified against the current code and deferring this one. `--filter` / `--filter-prod` are stored on `Config` for parity but are **not yet consumed by any recursive command** — `run -r` and `exec -r` also iterate every workspace project, and the CLI crate doesn't even depend on `pacquet-workspace-projects-filter`. The `Config::filter` doc comment documents this "stored now, consumed as the architecture lands" posture.

Making `pack -r` uniquely honor selectors would diverge from its sibling recursive commands. Wiring `filter_workspace_projects` into the shared recursive selection path is a separate cross-cutting change that should land for `run` / `exec` / `pack` together. Leaving this thread open to track that follow-up.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pacquet/crates/pack/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,668 @@
...
+        // Upstream globs the license with `onlyFiles` (fast-glob's
+        // default), so a directory named `LICENSE` is not a match.
+        // `fs::metadata` follows symlinks, mirroring fast-glob's default
+        // symlink resolution, so a symlinked license still counts while a
+        // directory (or a symlink to one) is skipped — the latter would
+        // otherwise fail the later read/size pass with "Is a directory".
+        if std::fs::metadata(entry.path()).is_ok_and(|metadata| metadata.is_file()) {
+            files_map.insert(format!("package/{name}"), workspace_dir.join(&name));
+        }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pacquet/crates/pack/src/lib.rs:629]: Fixed in 44136c3784 — the workspace LICENSE injection now uses `DirEntry::file_type()` (which does **not** follow symlinks) and requires `is_file()`, so a symlinked `LICENSE` is skipped rather than having its target's bytes packed. This matches the symlink-skipping `read_readme_file` in `exportable-manifest`, which mirrors pnpm's deliberately hardened `readReadmeFile` (`entry.isFile()` with the same "could leak its target's contents" rationale). Added a unix regression test (`symlinked_workspace_license_is_not_injected`).

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pacquet/crates/pack/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,668 @@
...
+    if !opts.ignore_scripts {
+        run_scripts_if_present::<Reporter>(opts, &["prepack", "prepare"], &entry_manifest)?;
+    }
+
+    // The publish directory may differ from the project root when
+    // `publishConfig.directory` redirects packing at a build output.
+    let dir = match publish_config_directory(&entry_manifest) {
+        Some(relative) => opts.dir.join(relative),
+        None => opts.dir.clone(),
+    };
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pacquet/crates/pack/src/lib.rs:232]: Declining this one — it would diverge from pnpm. Upstream's `pack.ts` joins `publishConfig.directory` onto the project dir with no containment check ([pack.ts#L218-L220](https://github.com/pnpm/pnpm/blob/54c5c0e028/pnpm11/releasing/commands/src/publish/pack.ts#L218-L220)):

```ts
const dir = entryManifest.publishConfig?.directory
  ? path.join(opts.dir, entryManifest.publishConfig.directory)
  : opts.dir
```

pacquet mirrors that exactly. The join *is* the `publishConfig.directory` feature, which legitimately points at a build-output directory; pnpm intentionally does not restrict it, and a containment guard would reject valid manifests while breaking parity (the cardinal rule in `pacquet/AGENTS.md`).

On threat model: `pack` operates on the **invoking user's own project** — `publishConfig.directory` lives in the root `package.json` they author and are packing on their own machine, so it isn't a remote-attacker-controlled trust boundary the way a downloaded dependency's manifest is (cf. the git-fetcher `bundleDependencies` escape checks, which guard genuinely untrusted input). If pnpm adds containment validation upstream, pacquet should follow then.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pacquet/crates/exportable-manifest/src/transform/tests.rs [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    let err = run(json!({ "name": "", "version": "1.0.0" })).unwrap_err();
+    assert_eq!(err, TransformError::MissingRequiredField { field: "name" });
+}
+
+#[test]
+fn falsy_version_is_rejected() {
+    for falsy in [json!(0), json!(false), json!(null)] {
+        let err = run(json!({ "name": "foo", "version": falsy })).unwrap_err();
+        assert_eq!(err, TransformError::MissingRequiredField { field: "version" });
+    }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pacquet/crates/exportable-manifest/src/transform/tests.rs:33]: Done in 44136c3784 — added `json!("")` to the falsy `version` set, so the empty-string `version` path is now pinned alongside `name`.

---
Written by an agent (Claude Code, claude-opus-4-8).

**Comment by zkochan** [comment]
**Re: Qodo finding #3 — "Tarball built in memory" (`tarball.rs`)**

Thanks — valid observation, tracking it as a follow-up rather than fixing it in this PR.

`pack` is not on the hot-path list the [REVIEW_GUIDE](https://github.com/pnpm/pnpm/blob/main/REVIEW_GUIDE.md) calls out (install / add / update / remove / resolution / fetching / linking / lockfile / store / pnpr), so the performance bar here is lower, and for typical packages the in-memory build is a few MB. Streaming to a `Write` would mean evolving the `Sys` capability seam from `FsWrite(path, bytes)` to a streaming interface (`create_file(path) -> impl Write`) and threading file handles through `build_tarball` and the size pass — a larger, self-contained refactor that's better as its own change than bundled into this command port. Filing it as a follow-up.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pacquet/crates/pack/src/lib.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (zkochan) [reply pacquet/crates/pack/src/lib.rs:232]: Fixed in e866c436a1 — `Host::write` now writes a sibling temp file, fsyncs it, then renames it over the target. The rename **replaces** a symlink at the output path instead of following it (so a repo-controlled symlink can no longer redirect the write to clobber an arbitrary file), and the write is atomic (no truncated `.tgz` on a crash mid-write). This mirrors the `write-file-atomic` temp-file + rename pattern `pacquet-package-manifest` already uses for `package.json` ([#12642](https://github.com/pnpm/pnpm/pull/12642)). The mode is preserved when overwriting an existing file (otherwise 0o644) so the archive isn't left owner-only.

The DI seam is unchanged — only the production `Host` impl became safe-by-default; fakes can still inject `PermissionDenied` / `ENOSPC`. Added a unix regression test (`symlinked_output_path_is_not_followed`) asserting a symlinked output path is replaced and its target left untouched.

Note: upstream pnpm writes the tarball with a plain `fs.createWriteStream` (non-atomic, follows symlinks), so this hardens pacquet slightly beyond pnpm — but the final file is identical for normal use, and it's consistent with pacquet's recent atomic-write direction.

---
Written by an agent (Claude Code, claude-opus-4-8).

[the remark this answers is not in the document]
Comment (zkochan) [reply pacquet/crates/pack/src/lib.rs:232]: Correction: my immediately-preceding reply (about atomic tarball writes) was posted on this thread by mistake — it belongs to the `Host::write` discussion in `capabilities.rs`, where I've now reposted it. This `publishConfig.directory` thread stands as **declined for pnpm parity** per my earlier reply above.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pacquet/crates/pack/src/capabilities.rs [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+impl FsCreateDirAll for Host {
+    fn create_dir_all(path: &Path) -> io::Result<()> {
+        std::fs::create_dir_all(path)
+    }
+}
+
+impl FsWrite for Host {
+    fn write(path: &Path, bytes: &[u8]) -> io::Result<()> {
+        std::fs::write(path, bytes)
+    }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pacquet/crates/pack/src/capabilities.rs]: Fixed in e866c436a1 — `Host::write` now writes a sibling temp file, fsyncs it, then renames it over the target. The rename **replaces** a symlink at the output path instead of following it (so a repo-controlled symlink can no longer redirect the write to clobber an arbitrary file), and the write is atomic (no truncated `.tgz` on a crash mid-write). This mirrors the `write-file-atomic` temp-file + rename pattern `pacquet-package-manifest` already uses for `package.json` ([#12642](https://github.com/pnpm/pnpm/pull/12642)). The mode is preserved when overwriting an existing file (otherwise 0o644) so the archive isn't left owner-only.

The DI seam is unchanged — only the production `Host` impl became safe-by-default; fakes can still inject `PermissionDenied` / `ENOSPC`. Added a unix regression test (`symlinked_output_path_is_not_followed`) asserting a symlinked output path is replaced and its target left untouched.

Note: upstream pnpm writes the tarball with a plain `fs.createWriteStream` (non-atomic, follows symlinks), so this hardens pacquet slightly beyond pnpm — but the final file is identical for normal use, and it's consistent with pacquet's recent atomic-write direction.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pacquet/crates/pack/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,668 @@
...
+/// or the default `<name>-<version>.tgz`. `--out` and
+/// `--pack-destination` are mutually exclusive.
+fn resolve_output(
+    opts: &PackOptions,
+    normalized_name: &str,
+    version: &str,
+) -> Result<(String, Option<String>), PackError> {
+    let Some(out) = &opts.out else {
+        return Ok((format!("{normalized_name}-{version}.tgz"), opts.pack_destination.clone()));
+    };
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pacquet/crates/pack/src/lib.rs:528]: Fixed in 384d827c00 — `pacquet_pack::api` now rejects a `version` containing `/` or `\\` with a new `PackError::InvalidPackageVersion` (`ERR_PNPM_INVALID_PACKAGE_VERSION`), mirroring the existing `name` validation right above it. A real semver version never contains a separator, so this only rejects the traversal case (`version: "../../evil"`) before it reaches the `<name>-<version>.tgz` join. Added `version_with_path_separator_is_rejected` covering `../../evil`, `1.0.0/../../evil`, and the backslash variant.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pacquet/crates/pack/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,668 @@
...
+    let pack_destination =
+        if parent.is_empty() { opts.pack_destination.clone() } else { Some(parent) };
+    Ok((tarball_name, pack_destination))
+}
+
+/// Map each packed path to `package/<path>` → absolute source, in
+/// packlist order.
+fn build_files_map(dir: &Path, files: &[String]) -> indexmap::IndexMap<String, PathBuf> {
+    files.iter().map(|file| (format!("package/{file}"), dir.join(file))).collect()
+}
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pacquet/crates/pack/src/lib.rs:553]: Fixed in 384d827c00 — the `main` / `bin` force-include in `pacquet-fs-packlist` now validates the normalized field with two new guards before inserting:

- `is_contained_field_path`: requires only `Component::Normal` parts (no `..`, root, or drive/UNC prefix) and no backslash, so `../secret` or a Windows absolute/UNC path can no longer escape the package root.
- `is_regular_file`: uses `symlink_metadata` instead of `is_file()`, so a `main`/`bin` pointing at a symlink to an out-of-tree file is not force-included.

This matters for the git fetcher too, which shares this crate and imports untrusted packages. A real package's `main`/`bin` is always an in-package relative regular file, so there's no parity cost. Added `escaping_main_and_bin_fields_are_not_force_included` ( `main`/`bin` = `../secret.js` with the target present, asserting it's not packed).

---
Written by an agent (Claude Code, claude-opus-4-8).

**Comment by zkochan** [comment]
**Disposition of the remaining findings in Qodo's review summary** (these were summary-only, with no inline threads):

| # | Finding | Disposition |
|---|---------|-------------|
| 1 | unsafe_perm hardcoded true | **Fixed** (`67bcd63eaa`) — `PackOptions.unsafe_perm` is now threaded from `Config.unsafe_perm` instead of hardcoded `true`, so lifecycle scripts honor the same policy/`TMPDIR` isolation as pnpm. |
| 7 | Unbounded README read | **Declined (parity)** — pnpm's `readReadmeFile` reads the whole file with no size cap either; `embed_readme` is also always-false today. |
| 8 | Recursive flag validation bypass | **Fixed** (`878dc37140`) — recursive pack now rejects `--out` + `--pack-destination` up front, matching single-project pack. |
| 9 | Terminal escape injection | **Fixed** (`67bcd63eaa`) — text output strips control characters from manifest/filesystem-derived paths; JSON output left as data. |
| 10 | Invalid `--out` filename accepted | **Fixed** (`67bcd63eaa`) — `--out .`/`..`/`""` is rejected with `ERR_PNPM_INVALID_OPTION` instead of a confusing later write error. |
| 11 | Alloc-heavy contents sorting | **Fixed** (`67bcd63eaa`) — one cached lowercase key per item instead of `to_lowercase` on both sides per comparison. |
| 12 | Quadratic bin mode check | **Fixed** (`67bcd63eaa`) — bin executability is checked against a `HashSet` instead of a linear scan. |
| 13 | Bin normalization assumes string | **Declined (unreachable)** — `pack` rejects a non-string `name` with `PackageNameNotFound` before `transform` runs; latent only for a future `publish`. |
| 14 | Tarball built in memory | **Fixed** (`67bcd63eaa`) — the gzip+tar archive now streams into the atomic temp file instead of buffering the whole `Vec`; per-file reads stay in memory, matching pnpm's `readFileSync`. |

(#2/#3/#4/#6 were resolved via their inline threads; #5 `publishConfig.directory` was declined for pnpm parity in its inline thread.)

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pacquet/crates/pack/src/capabilities.rs [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+        // A `NamedTempFile` is created 0o600. Match what a plain
+        // `fs::write` would leave: preserve an existing tarball's mode when
+        // overwriting, otherwise widen to 0o644 so the archive isn't
+        // owner-only.
+        #[cfg(unix)]
+        {
+            use std::os::unix::fs::PermissionsExt;
+            let mode = std::fs::metadata(path)
+                .map_or(0o644, |metadata| metadata.permissions().mode() & 0o777);
+            tmp.as_file().set_permissions(std::fs::Permissions::from_mode(mode))?;
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pacquet/crates/pack/src/capabilities.rs]: Fixed in 676729d382 — the mode preservation now uses `std::fs::symlink_metadata` and only inherits the mode from an existing **regular file** (`.filter(Metadata::is_file)`), so a symlink at the output path (to a directory or anything else) no longer donates its mode; it falls back to 0o644. I also moved `sync_all()` to after `set_permissions()` so the content and the final mode are flushed together before the rename.

---
Written by an agent (Claude Code, claude-opus-4-8).

File: pacquet/crates/fs-packlist/src/lib.rs [hunk 15]
```diff
@@ -554,6 +578,28 @@ fn normalize_field_path(path: &str) -> String {
...
+            .components()
+            .all(|component| matches!(component, Component::Normal(_)))
+}
+
+/// Whether `path` is a regular file without following a final symlink, so
+/// a `main` / `bin` pointing at a symlink to a file outside the package
+/// is not force-included (which would leak the target's bytes).
+fn is_regular_file(path: &Path) -> bool {
+    fs::symlink_metadata(path).is_ok_and(|metadata| metadata.file_type().is_file())
+}
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pacquet/crates/fs-packlist/src/lib.rs]: Fixed in 676729d382 — replaced the final-entry `symlink_metadata` check with `is_regular_file_within(root, candidate)`, which canonicalizes the candidate (following **all** symlinks, including an intermediate `subdir -> /outside`) and requires the resolved path to be strictly inside the canonical package root and a regular file. `is_contained_field_path` stays as a cheap lexical pre-filter. This also improves parity with npm-packlist: a `main`/`bin` symlink that stays **inside** the package is now followed and included (it was being skipped before), while any escape fails closed. Added `main_resolving_through_a_symlinked_dir_is_not_force_included` covering the intermediate-symlink case.

---
Written by an agent (Claude Code, claude-opus-4-8).

**Review by zkochan — APPROVED** [review]
